java-checks/src/main/java/org/sonar/java/checks/VirtualThreadNotSynchronizedCheck.java

```diff
@@ -0,0 +1,229 @@
+        .addParametersMatcher("java.lang.Runnable").build(),
+      MethodMatchers.create()
+        .ofTypes("java.lang.Thread")
+        .names("startVirtualThread")
+        .addParametersMatcher("java.lang.Runnable").build(),
+      MethodMatchers.create()
+        .ofSubTypes("java.util.concurrent.ExecutorService")
+        .names("execute", "submit")
+        .withAnyParameters().build()
+    );
+
+    private static final MethodMatchers EXECUTOR_BUILDER_SERVICE_WITH_VIRTUAL_TASKS_METHOD = MethodMatchers.create()
+      .ofTypes("java.util.concurrent.Executors")
+      .names("newVirtualThreadPerTaskExecutor")
+      .withAnyParameters().build();
+
+    public final List<MethodInvocationTree> invocations = new ArrayList<>();
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      super.visitMethodInvocation(tree);
+      if (VIRTUAL_THREAD_BUILDER_METHODS.matches(tree) && isRunnableInVirtualThread(tree)) {
+        invocations.add(tree);
+      }
+    }
+
+    private static boolean isRunnableInVirtualThread(MethodInvocationTree tree) {
+      return switch (tree.methodSymbol().name()) {
+        case "start", "unstarted" -> isCallToOfVirtual(tree);
+        case "execute", "submit" -> isCallToExecutorServiceWithVirtualTasks(tree);
+        default -> true;
+      };
+    }
+
+    private static boolean isCallToOfVirtual(MethodInvocationTree tree) {
+      var callSiteExpression = ((MemberSelectExpressionTree) tree.methodSelect()).expression();
+      if (callSiteExpression.symbolType().is(OF_VIRTUAL)) {
+        return true;
+      }
+
+      // Cover the case that call site is not type OfVirtual, but initialized with OfVirtual
+      return getInitializerExpression(callSiteExpression)
+        .stream()
+        .anyMatch(it -> it.symbolType().is(OF_VIRTUAL));
+    }
+
+    private static boolean isCallToExecutorServiceWithVirtualTasks(MethodInvocationTree tree) {
+      var callSiteExpression = ((MemberSelectExpressionTree) tree.methodSelect()).expression();
+      if (isCallToExecutorServiceBuilderWithVirtualTasks(callSiteExpression)) {
+        return true;
+      }
+
+      return getInitializerExpression(callSiteExpression)
+        .stream()
+        .anyMatch(RunnablesToCheckCollector::isCallToExecutorServiceBuilderWithVirtualTasks);
+    }
+
+    private static Optional<ExpressionTree> getInitializerExpression(ExpressionTree expression) {
+      return expression instanceof IdentifierTree identifier && identifier.symbol().declaration() instanceof VariableTree variableTree ?
+        Optional.ofNullable(variableTree.initializer()) : Optional.empty();
+    }
+
+    private static boolean isCallToExecutorServiceBuilderWithVirtualTasks(ExpressionTree expression) {
+      return expression instanceof MethodInvocationTree mit && EXECUTOR_BUILDER_SERVICE_WITH_VIRTUAL_TASKS_METHOD.matches(mit);
+    }
+  }
+
+  private static class TreesWithSynchronizedCodeCollector extends BaseTreeVisitor {
```

Comment:
If we don't scan the whole CompilationUnit anymore, I think there might be a way for this class to return directly an `Optional<Tree>` with the secondary location whenever synchronized code is spotted, and it could probably be more stateless if it could do so. 

So that in the end you could change the `onMethodInvocationFound` that I suggested earlier with something like:
```
  private static final TreesWithSynchronizedCodeCollector syncCollector = new TreesWithSynchronizedCodeCollector();
  @Override
  protected void onMethodInvocationFound(MethodInvocationTree mit) {
    if(isRunnableInVirtualThread(mit)) {
      syncCollector.isInvokingSynchronizedCode(mit.arguments().get(0)).ifPresent(secondaryLocation -> reportIssue(mit, secondaryLocation));
    }
  }
```